### PR TITLE
523: Use CAS2 with Nomis credentials

### PR DESCRIPTION
**Pull request by edavey**
## NB: BEFORE MERGING WE WILL:
-  ✅ get https://github.com/ministryofjustice/hmpps-approved-premises-api/pull/1009 merged and refactor the new `/cas2-people/search` endpoint added in this PR to use the separate `cas2-api.yml` OpenAPI spec and take the desired cas2 namespace form of `/cas2/people/search`
- ✅ co-ordinate a deployment of the UI repo so that the CAS2 front end uses this new endpoint for searching for people (see https://github.com/ministryofjustice/hmpps-community-accommodation-tier-2-ui/pull/246/)

## Overview

In this large PR, building on the earlier [PR 984: Extract Cas2Application entity from JPA inheritance](https://github.com/ministryofjustice/hmpps-approved-premises-api/pull/984) we make changes to allow:

- a user to authenticate with Nomis creds
- our service to store a representation of that user
- a `Cas2Application` to be associated with a Nomis user
- authorisation (principally for LAO) to be decoupled from Delius

#### New entities
- `Cas2ApplicationEntity` - This now lives outside of the JPA "joined" inheritance system. Everything is managed by this entity and is represented in the `cas_2_applications` table.

- `NomisUserEntity` - Similar to the existing (Delius) `UserEntity` this entity handles the CAS API's internal representation of known Nomis users. 

#### New API Client
- `NomisUserRolesApiClient` - This new client fetches information about the logged in Nomis user from the [Nomis-User-Roles API](https://github.com/ministryofjustice/nomis-user-roles-api/). This is also [now included in our AP-tools repo](https://github.com/ministryofjustice/hmpps-approved-premises-tools/pull/42) to allow developers to log in locally with Nomis credentials.

#### New model
- `NomisUserDetail` - An ephemeral representation of the JSON payload received from the Nomis-User-Roles API. Similar to `StaffUserDetails` on the Delius side.

#### New services
Main namespace:
- `NomisUserService` - supplies a logged in Nomis users' details from our db, sourced from Nomis-User-Roles API where unknown.

CAS2 namespace:
- `Cas2-ApplicationService` - similar to existing `ApplicationService` but without reliance on Delius user for LAO checks. The service will be extended to limit access based on Nomis roles and/or Nomis caseload
- `Cas2-OffenderService` - again similar to the existing OffenderService, but without LAO checks at Delius. And again this will be extended for the Nomis way of controlling access to offenders
- `Cas2-UserAccessService` - minimal authorisation class, currently implementing only `userCanViewApplication(user: NomisUserEntity, application: Cas2ApplicationEntity)` allowing access to only an application's author. This will be developed to grant permissions to the coming Nomis user roles (e.g. `CAS2_READ`, `CAS2_ADMIN`)

## IMPORTANT note on authentication

This PR proposes to change two shared classes in our authentication code:

### 1. `HttpAuthService`

Adding a `getNomisPrincipalOrThrow()` function to allow CAS2 to verify that the logged in user originates from the prison side:

```kotlin
fun getNomisPrincipalOrThrow(): AuthAwareAuthenticationToken {  
  val principal = SecurityContextHolder.getContext().authentication as AuthAwareAuthenticationToken  
  
  if (principal.token.claims["auth_source"] != "nomis") {  
    throw ForbiddenProblem()  
  }  
  
  return principal  
}
```

We will probably soon need some thing more flexible such as:

```kotlin
fun isValidCas2AuthSource(authSource: Any?): Boolean {
 return listOf("delius", "nomis", "nacro").contains(authSource)
}
```

But that can follow on as needed.

### 2. `OAuth2ResourceServerSecurityConfiguration`

We use `HttpServletRequest`'s `authorizeHttpRequests` mechanism
to define a `securityFilterChain` which controls access to
urls based on Roles (aka Authorities).

Reviewing the [Spring Security docs][spring_security_docs],
I feel like we ought to define 3 [custom `RequestMatcher`s][custom_request_matchers]

 e.g.

- `Cas1Access`
- `Cas2Access`
- `Cas3Access`

which I think would look something like the example given in the above docs :

```kotlin
val cas2Access: RequestMatcher = { (request) ->
  request.getServletPath().matches("cas2")
  // ...other conditions
}

http {
  authorizeHttpRequests {
    authorize(cas1Access, hasRole("PROBATION"))
    authorize(cas2Access, hasRole("PRISON"))
    authorize(cas3Access, hasRole("PROBATION"))
  }
}
```

I would hope that it would be much easier to unit test these
custom matchers than it is to unit test the entire
`OAuth2ResourceServerSecurityConfiguration.securityFilterChain()`
function. I've tried to test that but failed to mock out the
`SpringSecurityContext` successfully...

However, the [docs for our version 5 of Spring Security][spring_5_docs]
don't seem to mention the custom matchers and I encountered
errors attempting an implementation. So maybe we would need
to upgrade to v6 of Spring Security?

[spring_security_docs]:
https://docs.spring.io/spring-security/reference/servlet/authorization/authorize-http-requests.html

[custom_request_matchers]:
https://docs.spring.io/spring-security/reference/servlet/authorization/authorize-http-requests.html#match-by-custom

[spring_5_docs]:
https://docs.spring.io/spring-security/reference/5.8/servlet/authorization/authorize-http-requests.html

### CODE REVIEW

**Comment by edavey** [comment]
[Introduce NomisUser, associated with Cas2Application](https://trello.com/c/r2JKr0hy/523-introduce-nomisuser-associated-with-cas2application)

**Review by tahb — COMMENTED** [review]
[The changes to authorisation](https://github.com/ministryofjustice/hmpps-approved-premises-api/pull/991/commits/234e9de1c02c5206b3e9180138c32fce4c712dcd) seem reasonable to me. From reading through the Spring docs this does seem the way to do it.

Afraid this commit has too much in it for me to feel confident in approving the PR as a whole.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/config/OAuth2ResourceServerSecurityConfiguration.kt [hunk 1]
```diff
@@ -56,6 +56,7 @@ class OAuth2ResourceServerSecurityConfiguration {
         authorize(HttpMethod.DELETE, "/internal/premises/*", permitAll)
         authorize(HttpMethod.DELETE, "/internal/room/*", permitAll)
         authorize(HttpMethod.GET, "/events/**", hasAuthority("ROLE_APPROVED_PREMISES_EVENTS"))
+        authorize("/cas2*/**", hasRole("PRISON"))
```
[the remark this answers is not in the document]
Comment (edavey) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/config/OAuth2ResourceServerSecurityConfiguration.kt:61]: thks @tahb yes sorry I should have been clear about this alternate syntax. I believe that using `hasRole()` is the more idiomatic Spring syntax and that we should move to it -- but not right now!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas2/OffenderService.kt [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+      is ClientResult.Failure -> {
+        logFailedResponse(inmateDetailResponse)
+        null
+      }
+    }
+
+    return AuthorisableActionResult.Success(inmateDetail)
+  }
+
+  fun getOffenderByCrn(crn: String): AuthorisableActionResult<OffenderDetailSummary> {
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas2/OffenderService.kt:93]: What's the thinking behind namespacing this in CAS2?

**Review by pezholio — APPROVED** [review]
LGTM - A minor worry about there being a bit of duplication between services, but given the time constraints I think this is fine.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/config/OAuth2ResourceServerSecurityConfiguration.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (tahb) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/config/OAuth2ResourceServerSecurityConfiguration.kt]: 👍 One way to join this up might be to put a quick refactor PR in first that only changes the existing usage to hasAuthority so it's the way we want it to be.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas2/OffenderService.kt (hunk 2 above)
Comment (edavey) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas2/OffenderService.kt:93]: CAS2 needs to be disentangle itself from Delius users and so we've reimplemented a number of functions which we need but without their Delius connection -- which is usually in the form of a test for LAO access against Delius' `offenders/crn/{crn}/userAccess` endpoint:

- `getOffenderbyCrn(crn: String, userDistinguishedName: String, ignoreLao: Boolean = false)`
- `getInfoForPerson(crn: String, deliusUsername: String, ignoreLao: Boolean)`
- `getRiskByCrn(crn: String, jwt: String, userDistinguishedName: String)` In this function we also skip the upstream API calls to fetch Mappa, Tier and Flag info which is not needed

We do, however, continue to use the original `OffenderService` where possible, e.g. to get OASys information via:

- `OffenderService.getOASysOffenceDetails(crn)`
- `OffenderService.getOASysRiskToTheIndividual(crn)`
- `OffenderService.getOASysRoshSummary(crn)`

We bring the original OffenderService into our CAS2 PeopleController with:

`import uk.gov.justice.digital.hmpps.approvedpremisesapi.service.OffenderService as OASysOffenderService`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/NomisUserEntity.kt [hunk 3]
```diff
@@ -26,6 +28,9 @@ data class NomisUserEntity(
   var isActive: Boolean,
   var email: String?,
 
+  @CreationTimestamp
```
[the remark this answers is not in the document]
Comment (edavey) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/NomisUserEntity.kt:31]: yes, there's also a `@UpdateTimestamp` annotation which we could use...

**Comment by edavey** [comment]
> LGTM - A minor worry about there being a bit of duplication between services, but given the time constraints I think this is fine.

I think as time goes on we'll be able to make more common, shared behaviour and reduce the lines of code.

File: src/main/resources/db/migration/all/20230925133015__create_table_for_nomis_users.sql [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    updated_at      TIMESTAMP WITH TIME ZONE,
+
+    PRIMARY KEY (id),
+    UNIQUE (nomis_username)
+);
+
+-- truncate pre-existing cas2 applications
+--   we only have a 'dev' environment in which all applications are already truncated:
+--   see db/migration/dev+test/R__3_clear_applications.sql
+TRUNCATE TABLE cas_2_applications CASCADE;
```
[the remark this answers is not in the document]
Comment (edavey) [reply src/main/resources/db/migration/all/20230925133015__create_table_for_nomis_users.sql]: I've remove the truncation from the migration

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/config/OAuth2ResourceServerSecurityConfiguration.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (edavey) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/config/OAuth2ResourceServerSecurityConfiguration.kt]: thanks. I've changed them both to use the existing `hasAuthority()` function for now as that's clearer. I would rather put off any more refactoring till after this is merged!

**Review by Harriethw — APPROVED** [review]
LGTM, thank you for documenting and explaining this mammoth work so excellently.

**Pull request merged**